baybe/surrogates/gaussian_process/core.py

```diff
@@ -146,3 +147,11 @@ def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> No
             botorch.optim.fit.fit_gpytorch_mll_torch(mll, step_limit=200)
         else:
             botorch.fit.fit_gpytorch_mll(mll)
+
+    def __str__(self) -> str:
+        fields = [
+            f"Joint posterior: {self.joint_posterior}",
+            f"Supports Transfer Learning: {self.supports_transfer_learning}",
+            f"Kernel factory: {self.kernel_factory}",
+        ]
+        return "\n".join([self.__class__.__name__] + [indent(f) for f in fields])
```

Comment:
I haven't yet grasped the logic how you nest the different sub-entries of the hierarchy. In particular, this line seems to be present more less in every class, but sometimes it starts with a `"\n"` and sometimes with a `"\n\n"`. Looks like stuff is hard-coded / not yet generalized?

Reply:
For me, that was a decision that I made based on the visual appearance: For most classes, using a single `\n` looked good, for some it looked to cluttered (mainly for `campaign` iirc) which is why I used `\n\n`. If you prefer consistency over visual appearance, we can go for either variant everywhere, I do not have a strong opinion here.

Reply:
So if you ask me, I'd avoid hard-coding stuff altogether. Right now, you are about to customize stuff for our current object hierarchy (e.g. having `Campaign` in mind as highest-level object). But things are changing all the time and also will continue to change in the future. So I'd prefer if we have a completely generic printing approach that works regardless of the content/context/current class layout. 

Have a look, for example, how torch prints stuff:
<img width="453" alt="image" src="https://github.com/user-attachments/assets/7d22ae52-e9fe-4d52-9b62-81fc10310c94">
Of course we don't need to mirror this 1:1, but the advantage are clearly:
* Works regardless of object hierarchy and content / you don't need to thing about spacing
* The closing parenthesis also shows you where an object ends.

And these are the two features I'd also like to have ideally. Instead of all these parenthesis, perhaps we can use a visually appealing signal for "start" and "end". For instance a "T" and an inverted "T", or something similar that "spans an implicit line"? 

This is just a first idea:
<img width="440" alt="image" src="https://github.com/user-attachments/assets/a31cd323-be7f-46ce-88a0-d869dc2f47bc">
In any case, I'd simply outsource this to a printing function (similar like the `indent` thingy you added) which would allow us to easily play around / modify later on

Reply:
Have to say that I do not like that "inverted T" approach - for me, that looks way too cluttered. Also, I am not a big fan of the botorch approach.
The question is: Do we want our printing to print actually *everything* like in the examples you provide or do we want to have a *pretty printing*? Originally, the goal was to have a sensible, reasonable and human readable version that explicitly does not list absolutely everything - hence the approach to have something manually, where we decide what we want to print and how. Also summoning @Scienfitz here for his opinion.

Reply:
Not saying that we need to print everything, I'd be happy with a setting where we manually select stuff to be shown. But that decision is orthogonal to decision *how* stuff is being printed. All I'm saying at the moment: Rim has chosen a hard-coded way of producing indents / spacing and now you are already doing the second iteration (which I appreciate) of fixing this part. When we now custom tailor the output again to the current class layout, I'm pretty sure we'll run into the next fix/adjustment in the next few months.

So I'm not promoting the "T" approach here, but just one that is completely generic in terms of created visuals (i.e. no manual decisions where to put double spacing etc)

Reply:
Should we then maybe just use some already existing package like `pprint` and just have that print our dict/json representation? This will be safe, has no custom tailoring from our side and should work out of the box.

Reply:
@AdrianSosic How about this?
![image](https://github.com/user-attachments/assets/f20b53a9-b682-4741-a8bf-95330a3c03f0)

Reply:
Could also increase indent: 
![image](https://github.com/user-attachments/assets/b8aa29ba-1adc-4a72-824f-4779e10d8416)

Reply:
why adding these fillers? Dont see what it has to do with the hardcoding arg of Adrian

Reply:
So I personally like it (it's essentially my suggestions without the "T" line endings) because I think it improves readability quite a lot – probably it would be really hard to parse the indentation level without the fillers. But I don't want the guy deciding this, should be a team decision. 

(BTW: there a still a few things not perfectly aligned, like the target entry and the metadata info, but that's not the point of the discussion here)

@Scienfitz: what would be your preference? Same thing without lines?

Reply:
hmm imo looks pretty terrible, are you saying the indents are not enough for your eye? For me its more thane enough

Reply:
You know, I'm an old man here sitting in front of my screen with font size 36, barely able to comprehend what you write :D 

Jokes aside, as I said: for me it helps visual parsing (which is IMO the main point of deviating from the default str method), but I am completely OK with the other option if both of you prefer it – as long is we keep it general and don't need to manually tweak again as soon as we change something in our object hierarchy.

Reply:
Regarding the changes made here: This change was included by simply changing the `indent` function to always add a `|` in front, and has nothing to do with the generalization. The target stuff is fixed by now, thanks for spotting the meta data issue.
The main question that we need to ask ourselves here is: What do we actually want? The only thing that is chosen manually here is now the symbol used in the beginning and the number of spaces that we want to indent, but we can simply change any of this to whatever we want. We can go with `|` and 2 resp. 3 spaces for indentation (these are the screenshots above), we can get rid of the `|`, and so on. But I think this is something that we should discuss in person by just looking together at some variants and then just agree on one.

Reply:
@AdrianSosic @Scienfitz this is how it would look like in the documentaiton with an indent of 2 and the horizontal bar:
<img width="941" alt="image" src="https://github.com/user-attachments/assets/357bd311-f755-4192-991b-1497f2234504">
Is this acceptable for everyone?

Reply:
if it helps our eldest member to see clearly keep it, but it shall be noted to protocol that I think this looks very ugly now (talking about the filling with bars Ts or anything thats not whitespace)

Reply:
😄 can you you show the alternative without bars?

And still question: how come that there are still inconsistencies when everything is fully automatic now? Talking about:
* Sometimes colon, sometimes not. Compare first two lines
* Why is the horizontal indent not consistent? Compare vertical lines below `Meta Data` to the other lines
* Isn't the indentation level still wrong in some places? For example, shouldn't everything below `Search Space` have one more indent?